flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/ChangelogNormalizeSemanticTestPrograms.java

```diff
@@ -0,0 +1,236 @@
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink_t")
+                                    .addSchema(SINK_SCHEMA)
+                                    .enforceRetract()
+                                    .consumedValues(
+                                            "+I[one, 1, a]",
+                                            "+U[one, 2, bb]",
+                                            "+I[one, 4, aaaa]",
+                                            "+U[one, 4, aaaa]",
+                                            "+I[two, 1, d]",
+                                            "+U[two, 1, d]",
+                                            "-D[two, 1, d]")
+                                    .build())
+                    .runSql("INSERT INTO sink_t SELECT a, b, c FROM source_t WHERE a <> 'three'")
+                    .build();
+
+    static final TableTestProgram CHANGELOG_NORMALIZE =
+            TableTestProgram.of("changelog-normalize", "validates changelog normalize")
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("source_t")
+                                    .enforceRetract()
+                                    .addSchema(SOURCE_SCHEMA)
+                                    .producedValues(
+                                            Row.ofKind(RowKind.INSERT, "one", 1, "a"),
+                                            Row.ofKind(RowKind.UPDATE_AFTER, "one", 2, "bb"),
+                                            Row.ofKind(RowKind.UPDATE_AFTER, "three", 3, "ccc"),
+                                            Row.ofKind(RowKind.UPDATE_AFTER, "three", 3, "ccc"),
+                                            Row.ofKind(RowKind.INSERT, "one", 4, "aaaa"),
+                                            Row.ofKind(RowKind.UPDATE_AFTER, "one", 4, "aaaa"),
+                                            Row.ofKind(RowKind.DELETE, "three", 3, "ccc"))
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink_t")
+                                    .enforceRetract()
+                                    .addSchema(SINK_SCHEMA)
+                                    .consumedValues(
+                                            "+I[one, 1, a]",
+                                            "+U[one, 2, bb]",
+                                            "+U[three, 3, ccc]",
+                                            "+I[one, 4, aaaa]",
+                                            "+U[one, 4, aaaa]",
+                                            "+U[three, 3, ccc]",
+                                            "-D[three, 3, ccc]")
+                                    .build())
+                    .runSql("INSERT INTO sink_t SELECT a, b, c FROM source_t")
```

Comment:
test cases that are missing now:
- upsert source with partial deletes and UA, D. This corresponds to Kafka.